src/main.cairo

```diff
@@ -1,32 +1,231 @@
+    pricing_contract.write(pricing_address);
 
-    let (res) = balance.read();
-    balance.write(res + amount);
+    Proxy.initializer(admin);
     return ();
 }
 
+// @notice Get the status of renewals for a domain
+// @param domain Domain to get status of
+// @param user User to get status of
 @view
-func get_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (res: felt) {
-    let (res) = balance.read();
+func will_renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt, 
+    renewer: felt
+) -> (res: felt) {
+    let (res) = is_renewing.read(renewer, domain);
     return (res,);
 }
 
-@constructor
-func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
-    balance.write(0);
+// @notice Get the status of an upgrade vote
+// @param upgrade Upgrade to get status of
+// @param user User to get status of
+@view
+func has_voted_upgrade{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    user: felt,
+    upgrade_id: felt, 
+    implementation_hash: felt,
+) -> (res: felt) {
+    let (res) = voted_upgrade.read(user, upgrade_id, implementation_hash);
+    return (res,);
+}
+
+//
+// Externals
+//
+
+// @notice Activate or deactivate renewals for a domain
+// @param domain Domain to activate or deactivate renewals for
+// @dev callable by owner of domain
+@external
+func toggle_renewals{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt,
+) {
+    alloc_locals;
+    let (caller) = get_caller_address();
+    let (contract) = naming_contract.read();
+
+    let (prev_renew) = is_renewing.read(caller, domain);
+    is_renewing.write(caller, domain, 1 - prev_renew);
+
+    toggled_renewal.emit(domain, caller, 1 - prev_renew);
+
+    return ();
+} 
+
+// @notice Renew a domain
+// @param domain Domain to renew
+// @dev callable by anyone, but will only renew if renewals are activated for domain
+@external
+func renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    root_domain: felt, renewer: felt,
+) {
+    alloc_locals;
+    let (naming) = naming_contract.read(); 
+    let (can_renew) = is_renewing.read(renewer, root_domain);
+
+    with_attr error_message("Renewer has not activated renewals for this domain") {
+        assert_not_zero(can_renew);
+    }
+
+    let (expiry) = Naming.domain_to_expiry(naming, 1, cast(new(root_domain), felt*));
+    let (block_timestamp) = get_block_timestamp();
+    with_attr error_message("Domain is not set to expire within a month") {
+        assert_le(expiry, block_timestamp + 86400 * 30);
+    }
+
+    let (pricing) = pricing_contract.read();
+    let (erc20, renewal_price) = Pricing.compute_renew_price(pricing, root_domain, 365);
+
+    let (contract) = get_contract_address();
+    with_attr error_message("Error transferring tokens from renewer to contract") {
+        IERC20.transferFrom(erc20, renewer, contract, renewal_price);
```

Comment:
Shouldn't we transfer it to the `naming contract`? With that, only one contract collects every fees.

Reply:
OKK